presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java

```diff
@@ -222,22 +222,30 @@ else if (nameParts.size() == 2) {
         Map<String, String> requestHeaders = getRequestHeaders(servletRequest);
         TracerHandle tracerHandle = tracerProvider.getHandleGenerator().apply(requestHeaders);
 
-        if (isTracingEnabled()) {
+        TraceTokenFactory traceTokenFactory = new PrestoTraceTokenFactory();
+
+        if (traceTokenFactory.getTraceToken(clientInfo).isPresent()) {
             this.tracer = Optional.of(requireNonNull(tracerProvider.getNewTracer(tracerHandle), "tracer is null"));
-            traceToken = Optional.ofNullable(this.tracer.get().getTracerId());
+            traceToken = (traceTokenFactory.getTraceToken(clientInfo));
         }
         else {
```

Comment:
better to make this `else if (isTracingEnabled())` to reduce layers of nesting, and be less error prone.